What is a function?
-------------------

A function is a piece of code that performs a particular task.  Defining one allows you to create new "words" that will  do the job for you, encapsulating a complex task into a simple command.  


A function definition has to have several elements:
=============

1. The keyword "def", short for "define".
2. The name of the function.  This can be any word you choose, but it is best if it is a word that means or helps you remember the task that the function performs, and it should not be a word that is already in use (a part of python), or in use in some other part of your program.
3. Parentheses after the name.  
4. A colon
5. Indented underneath the line that names the function, it has to do something.



In [ ]:
def doSomething():
    print("I did something")
    
doSomething()

This was not a particularly exciting function, but it is a function that does something.

1. It has the keyword **def**.
2. It has a name that reminds me what the function does.  **doSomething**
3. It has parentheses after the name.
4. It has a colon after the parentheses.
5. Indented underneath the line that names the function, it does something. **it prints**


The final line of the cell above is not part of the function, it is the line that **calls** the function.  You call a function by saying its name with parentheses.  The program then does what the function tells it to.

Make a function that does something more useful
---------------------------------------------

Make a function that squares a number.  It should *print* the squared value of the input number

In [ ]:
def square(number):
    result = number*number
    print(result)
    
square(15)

Notice that this function for squaring still has all the required elements:

1. It has the keyword **def**.
2. It has a name that reminds me what the function does.  **square**
3. It has parentheses after the name.
4. It has a colon after the parentheses.
5. Indented underneath the line that names the function, it does something. **it computes a value and prints it**

What's new is that there is a word *inside* the parentheses, *number*, and when you call the function, there is an actual number inside the parentheses.

When you call the function and put the number inside the parentheses, you are sending that number into the function.  Inside the function, python puts that number into the variable in the matching spot in the parentheses -- so the word "number" now is a variable that holds a particular value, in this case, the number 15.  

Then the function squares the number by multiplying by itself, 

stores the result in the word "result", 

and then prints it.

Try calling the function a few more times with different numbers sent or "passed" into the function.

In [ ]:
square(5)
square(0)
square(-2)

What can you do in a function?
-----------------------------

So what all  can you do inside a function?  Anything you can do in the  rest of the program, but it's a good idea to try to limit every function to just one task -- that makes it easier to debug if something goes wrong.

There's one additional thing that functions can do that we need to discuss at this point -- when they are finished, they can "return" a value back to the main program.  When you call the function you can "pass" values to it, when it finishes, it can "return" one thing back.  This can allow you to use that returned value in the rest of the program.



In [ ]:
def square(number):
    result = number*number
    return result
    
sq15 = square(15)

It doesn't look like that code cell did anything, because it didn't print anything -- but it did do something.  It has now stored a value in the variable sq15.  What value?  Let's print it:


In [ ]:
print(sq15)

So the value that came from inside the function is now available for use outside the function.

Let's make another function that takes two numbers, multiplies them together, then outputs the result.



In [ ]:
def multiply(a,b):
    result = a*b
    return result

multiply(6,4)

When you pass values into the function like this, they are assigned to the holder variables inside the function in order.  So in that first example of calling the multiply function, inside the function the variable **a** ends up having the value **6** and **b** has the value **4**

This is more important when you will do different things with the variables, for example, a division function:

In [ ]:
def divide(a,b):
    result = a/b
    return result

print("6 divided by 3:  ", divide(6,3))
print("3 divided by 6:  ", divide(3,6))

Task 1
-----

Make a function that prints a number every time you call it.  It doesn't matter what the number is

In [ ]:
def print_a_number():
    print(4)

In [ ]:
print_a_number()

Task 2
------
Make a function that takes an input number and prints the result of dividing it by 3.



In [ ]:
def div_by_3(number):
    result = number/3
    print(result)

In [ ]:
div_by_3(7)

Task 3
------

Make a function that takes as input two numbers, then **returns** the sum of all positive integers *between* those two numbers.

For example, if the input was 2 and 5, the returned value should be **7**, because 3+4 = 7.  Notice that 2 and 5 are not included because they are not *between* 2 and 5. 

This function will have to combine all the extra elements of multiple input values, a function body that does something substantial, and returning a value.  It will also be an example of the accumulator pattern.

In [ ]:
def sum_between(a,b):
    total = 0
    for num in range(a+1,b):
        total = total + num
    return total

In [ ]:
sum_between(2, 5)

Task 4
------


I'll get you some new data to look at, in this case it is a data set of 1.6 million tweets harvested from Twitter.  The columns of the dataset are:

sentiment: a numerical score indicating the emotion inferred from the text of the tweet (0 = negative, 2=neutral, 4 =positive)  These scores were assigned by people reading the tweets.  The dataset was developed to aid in training algorithms to recognize the sentiment in human generated text.

id: a unique numeric identifier for the particular tweet

datetime:  the date and time the tweet was sent

query: a column you can ignore for now

user: the username of the account that sent the tweet

text:  the actual text of the sent tweet

In [1]:
import pandas as pd

In [34]:
df = pd.read_csv('../twitter.10.csv', header=None, encoding="ISO-8859-1")
df.columns = ['sentiment', 'id', 'datetime', 'query', 'user', 'text']

In [35]:
len(df.user.unique())

105907

There are 105907 unique users in this dataset.  Here are the first ten:


In [36]:
df['user'].head(10)

0         jacqqq_xo
1           mykerob
2           TukeGuy
3          kevinlcc
4           OStephy
5       singingbell
6    megsapatsfan12
7            Danlol
8          Wildaris
9       bagthoughts
Name: user, dtype: object

Make a function that takes as input the dataframe and the name of a single user.  The function should return the number of tweets sent by that user.

In [45]:
def count_tweets(df, user):
    count = 0
    for entry in df.index:
        if df.loc[entry]['user'] == user:
            count += 1
    return count
        

In [47]:
count_tweets(df, 'tweetpet')

297

The function you produce here is probably going to be really slow -- there are more elegant ways of accomplishing this task, but for now I want you to use the approach of writing your own function to practice that skill.  If your function doesn't seem to be working, please check in with me quickly so I can help determine if it is working but it's slow, or if it is in fact not working.

Once you get the function to work, you can continue to explore the dataset in other ways, or you can submit it as is.

In [43]:
maxtweets = 0
for user in df['user'].unique():
    usertweets = len(df[df['user'] == user])
    if usertweets > maxtweets:
        maxtweets = usertweets
        print(user, usertweets)

jacqqq_xo 1
TaraRahimi 7
paigeebaby 21
Dogbook 44
tweetpet 297


KeyboardInterrupt: 